### PR TITLE
observer: fix split error

### PR DESCRIPTION
**Pull request by BusyJay**
@ngaut @siddontang @coocood PTAL

### CODE REVIEW

**Comment by siddontang** [comment]
LGTM

File: src/raftstore/coprocessor/split_observer.rs [hunk 1]
```diff
@@ -12,59 +12,45 @@
...
-                                   key,
-                                   expected_len));
-            }
-            new_key = &key[..expected_len];
+        let table_prefix_len = table::TABLE_PREFIX.len() + table::ID_LEN;
+        if key[table_prefix_len..].starts_with(table::RECORD_PREFIX_SEP) {
+            // truncate to handle
+            key.truncate(table::PREFIX_LEN + table::ID_LEN);
+        } else if key[table_prefix_len..].starts_with(table::INDEX_PREFIX_SEP) {
+            // it's an index key, just remove the tailing version_id.
```
Comment (ngaut) [inline src/raftstore/coprocessor/split_observer.rs]: s/version_id/version

File: src/raftstore/coprocessor/split_observer.rs [hunk 2]
```diff
@@ -12,59 +12,45 @@
...
-        bytes::decode_bytes(split_key)
-            .map_err(|e| format!("invalid key {:?}: {}", split_key, e))
-            .map(|r| r.0)
-    }
-
-    fn on_split(&mut self, _: &mut ObserverContext, split: &mut SplitRequest) -> Result<()> {
-        let key = try!(self.extract_sql_key(split));
+        let mut key = split.get_split_key().to_vec();
 
         // format of a key is TABLE_PREFIX + table_id + TABLE_ROW_MARK + handle or
```
Comment (ngaut) [inline src/raftstore/coprocessor/split_observer.rs]: Please add version to format

File: src/raftstore/coprocessor/split_observer.rs [hunk 3]
```diff
@@ -136,32 +125,22 @@ mod test {
...
-               column_id: u64,
-               version_id: u64)
-               -> Vec<u8> {
-        let mut key = Vec::with_capacity(100);
-        key.write(TABLE_PREFIX).unwrap();
-        key.write_u64::<BigEndian>(table_id).unwrap();
-        key.write(mark).unwrap();
-        key.write_u64::<BigEndian>(row_id).unwrap();
+    fn new_row_key(table_id: i64, row_id: i64, column_id: u64, version_id: u64) -> Vec<u8> {
+        let mut buf = vec![0; 8];
```
Comment (ngaut) [inline src/raftstore/coprocessor/split_observer.rs]: s/8/ID_LEN  ?

**Comment by ngaut** [comment]
PTAL @disksing @coocood

File: src/raftstore/coprocessor/split_observer.rs [hunk 4]
```diff
@@ -12,59 +12,49 @@
...
+
+        // format of a key is TABLE_PREFIX + table_id + TABLE_ROW_MARK + handle + column_id
+        // + version or TABLE_PREFIX + table_id + TABLE_INDEX_MARK + index_id + values + version
+        // or meta_key + version
+
+        let table_prefix_len = table::TABLE_PREFIX.len() + table::ID_LEN;
+        if key.starts_with(table::TABLE_PREFIX) && key.len() > table::PREFIX_LEN + table::ID_LEN &&
+           key[table_prefix_len..].starts_with(table::RECORD_PREFIX_SEP) {
+            // row key, truncate to handle
+            key.truncate(table::PREFIX_LEN + table::ID_LEN);
```
Comment (coocood) [inline src/raftstore/coprocessor/split_observer.rs]: It's OK to truncate to a certain length for now, but in the future, we will use string handle, the prefix length will change.

We can append row key a column ID 0, so column key and row key would have the same length, then truncate suffix will ensure a row will not be split.

**Comment by coocood** [comment]
LGTM

File: src/raftstore/coprocessor/split_observer.rs [hunk 5]
```diff
@@ -12,59 +12,49 @@
...
-                                   expected_len));
-            }
-            new_key = &key[..expected_len];
+
+        // format of a key is TABLE_PREFIX + table_id + TABLE_ROW_MARK + handle + column_id
+        // + version or TABLE_PREFIX + table_id + TABLE_INDEX_MARK + index_id + values + version
+        // or meta_key + version
+
+        let table_prefix_len = table::TABLE_PREFIX.len() + table::ID_LEN;
+        if key.starts_with(table::TABLE_PREFIX) && key.len() > table::PREFIX_LEN + table::ID_LEN &&
```
Comment (disksing) [inline src/raftstore/coprocessor/split_observer.rs]: It's better to throw an error and cancel split when the key starts with TABLE_PREFIX but has a length shorter than it should be.

Comment (BusyJay) [inline src/raftstore/coprocessor/split_observer.rs]: It may be meta key.

**Comment by disksing** [comment]
LGTM

File: src/raftstore/coprocessor/split_observer.rs (hunk 5 above)
Comment (coocood) [inline src/raftstore/coprocessor/split_observer.rs]: meta key will not have TABLE_PREFIX.

**Pull request merged**